juju/client/connector.py

```diff
@@ -103,7 +103,7 @@ async def connect_controller(self, controller_name=None, specified_facades=None)
         if not controller_name:
             controller_name = self.jujudata.current_controller()
         if not controller_name:
-            raise JujuConnectionError('No current controller')
+            raise JujuConnectionError('No current controller. Is Juju bootstrapped?')
```

Comment:
I'm not a fan of `current_controller()` returning `None` and us interpreting it here. I think we should just raise this error at the spot and have it bubble up.

Reply:
I did mention the exception approach in the "Notes & Discussion", but here is why I leaned toward `None`:

The class we're calling on (`FileJujuData`) is an interface to the Juju files. We're asking it "what do the files say the current controller is?", and I think the correct answer is actually "the files say there isn't one".

It's not really a connection error in `FileJujuData`, but it does become a connection error in the client since there is no controller for us to connect to.

There might be a better argument for raising a `JujuError`, but is that even true? Is it really an error if this file doesn't exist, or does it just mean some things haven't been set up yet, and therefore some operations are restricted? The distinction is important, because this only becomes an error once you decide to do something that requires Juju to have a controller set.

Reply:
Well, I managed to beat your late response with an extremely delayed response! I was caught up in something that was a little more urgent.

> I'm glad you said that we're using an interface, the thing is the FileJujuData is not the interface, but the implementation of the interface

Yes, sorry, I didn't mean interface from a language contruct perspective, just that `FileJujuData`'s job is to be the intermediary between code and the file.

I think you've convinced me it's reasonable to classify this an error. I'll add a new error type and we can see how it looks. If you have more comments I'll do my best to iterate more quickly.

Reply:
Does https://github.com/juju/python-libjuju/pull/937/commits/747dc771710bd0295c4480c0927d07a144e39b17 make more sense?